File: S2.1.5.kt
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.util.Base64

fun decryptAES(key: ByteArray, ciphertext: ByteArray): String {
    val iv = ciphertext.sliceArray(0 until 16) // Extract IV from ciphertext
    val encrypted = ciphertext.sliceArray(16 until ciphertext.size)

    val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
    val secretKey = SecretKeySpec(key, "AES")
    val ivParams = IvParameterSpec(iv)

    cipher.init(Cipher.DECRYPT_MODE, secretKey, ivParams)
    val decryptedBytes = cipher.doFinal(encrypted)

    return String(decryptedBytes)
}

fun main() {
    val key = "32-byte-long-encryption-key".toByteArray() // Replace with your AES key
    val encryptedBase64 = "<Base64 Encoded Ciphertext from Go>".toByteArray()
    val encryptedBytes = Base64.getDecoder().decode(encryptedBase64)

    val decryptedMessage = decryptAES(key, encryptedBytes)
    println("Decrypted: $decryptedMessage")
}

File: S2.1.5.go
package main

import (
    "crypto/aes"
    "crypto/cipher"
    "crypto/rand"
    "encoding/base64"
    "fmt"
    "io"
)

func encryptAES(key, plaintext []byte) ([]byte, error) {
    block, err := aes.NewCipher(key)
    if err != nil {
        return nil, err
    }

    ciphertext := make([]byte, aes.BlockSize+len(plaintext))
    iv := ciphertext[:aes.BlockSize]
    if _, err := io.ReadFull(rand.Reader, iv); err != nil {
        return nil, err
    }

    mode := cipher.NewCBCEncrypter(block, iv)
    mode.CryptBlocks(ciphertext[aes.BlockSize:], plaintext)

    return ciphertext, nil
}

func main() {
    key := []byte("32-byte-long-encryption-key") // Replace with your AES key
    plaintext := []byte("Hello, AES encryption!")

    ciphertext, err := encryptAES(key, plaintext)
    if err != nil {
        fmt.Println("Error encrypting:", err)
        return
    }

    fmt.Println("Encrypted:", base64.StdEncoding.EncodeToString(ciphertext))
}
